apps/webapp/app/send/send-form.tsx

```diff
@@ -1,33 +1,57 @@
 'use client';
 
+import dynamic from 'next/dynamic';
+import { LoHi, uint8ArrayToBase64 } from 'penumbra-types';
 import { useEffect } from 'react';
 import { Button, Switch } from 'ui';
-import { FilledImage, InputBlock, InputToken } from '../../shared';
+import { useBalances } from '../../hooks/balances';
+import { FilledImage, InputBlock } from '../../shared';
 import { useStore } from '../../state';
 import { sendSelector } from '../../state/send';
-import { validateAmount, validateRecipient } from '../../utils';
-import { assets } from './constants';
+import { calculateBalance, validateAmount, validateRecipient } from '../../utils';
+const InputToken = dynamic(() => import('../../shared/input-token'), {
+  ssr: false,
+});
 
-export const SendForm = () => {
+export default function SendForm() {
   const {
     amount,
     asset,
     recipient,
     memo,
     hidden,
     validationErrors,
+    assetBalance,
     setAmount,
     setAsset,
     setRecipient,
     setMemo,
     setHidden,
+    setAssetBalance,
   } = useStore(sendSelector);
 
+  const { data, end } = useBalances(0);
+
   useEffect(() => {
-    // assign an asset when the page loads
-    setAsset(assets[0]!);
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, []);
+    if (!end) return;
+    const selectedAsset = data.find(
+      i =>
+        i.balance?.assetId?.inner &&
+        uint8ArrayToBase64(i.balance.assetId.inner) === asset.penumbraAssetId.inner,
+    );
+
+    if (!selectedAsset) {
+      setAssetBalance(0);
+      return;
+    }
+
+    const loHi: LoHi = {
+      lo: selectedAsset.balance?.amount?.lo ?? 0n,
+      hi: selectedAsset.balance?.amount?.hi ?? 0n,
+    };
+
+    setAssetBalance(calculateBalance(loHi, asset));
+  }, [data, end, asset, setAssetBalance]);
```

Comment:
This looks like redundant code to the other code block. Can you extract this to a hook so it's re-used?